test/topology_custom/test_replace.py

```diff
@@ -9,19 +9,60 @@
 import time
 from test.pylib.scylla_cluster import ReplaceConfig
 from test.pylib.manager_client import ManagerClient
-from test.topology.util import wait_for_token_ring_and_group0_consistency
+from test.topology.util import wait_for_token_ring_and_group0_consistency, wait_for_cql_and_get_hosts, wait_for
 import pytest
+import logging
+
+
+logger = logging.getLogger(__name__)
 
 
 @pytest.mark.asyncio
 async def test_replace_different_ip(manager: ManagerClient) -> None:
     """Replace an existing node with new node using a different IP address"""
     servers = await manager.servers_add(3, config={'failure_detector_timeout_in_ms': 2000})
+    logger.info(f"cluster started, servers {servers}")
+
+    logger.info(f"replacing server {servers[0]}")
     await manager.server_stop(servers[0].server_id)
-    replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = False, use_host_id = False)
-    await manager.server_add(replace_cfg)
+    replaced_server = servers[0]
+    replace_cfg = ReplaceConfig(replaced_id = replaced_server.server_id, reuse_ip_addr = False, use_host_id = False)
+    new_server = await manager.server_add(replace_cfg)
+    cql = manager.get_cql()
+    servers = await manager.running_servers()
+    all_ips = set([s.rpc_address for s in servers])
+    logger.info(f"new server {new_server} started, all ips {all_ips}, "
+                "waiting for token ring and group0 consistency")
     await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
 
+    for s in servers:
+        peers_to_see = all_ips - {s.rpc_address}
+
+        logger.info(f'waiting for cql and get hosts for {s}')
+        h = (await wait_for_cql_and_get_hosts(cql, [s], time.time() + 60))[0]
+
+        logger.info(f"waiting for {s} to see its peers {peers_to_see}")
+        async def check_peers_and_gossiper():
+            peers = set([r.peer for r in await cql.run_async("select peer from system.peers", host=h)])
+            remaining = peers_to_see - peers
+            if remaining:
+                logger.info(f"server {h} doesn't see its peers, all_ips {all_ips}, peers_to_see {peers_to_see}, remaining {remaining}, continue waiting")
+                return False
```

Comment:
Hm, but `wait_for` is not waiting for `False`, it's waiting for `None` (because the predicate may return a value once it the wait finishes -- and that value can be also a boolean type):
```python3
async def wait_for(
        pred: Callable[[], Awaitable[Optional[T]]],
        deadline: float,
        period: float = 1,
        before_retry: Optional[Callable[[], Any]] = None) -> T:
    while True:
        assert(time.time() < deadline), "Deadline exceeded, failing test."
        res = await pred()
        if res is not None:
            return res
        await asyncio.sleep(period)
        if before_retry:
            before_retry()
```

So the `wait_for` below will always return immediately

Reply:
fixed